src/Validation/Rules/Ip.php

```diff
@@ -0,0 +1,25 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tempest\Validation\Rules;
+
+use Attribute;
+use Tempest\Interface\Rule;
+
+#[Attribute]
+final readonly class Ip implements Rule
+{
+    public function isValid(mixed $value): bool
+    {
+        return (
```

Comment:
I think this nested conditional is a bit confusing; maybe rewrite it in several steps so that it's more clear to what it does. I'm ok with multiple `if` statements and early returns.